src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqProvider.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    internal abstract class AbstractConvertLinqProvider : CodeRefactoringProvider
+    {
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var semanticModel = await document.GetSemanticModelAsync().ConfigureAwait(false);
+
+            await CreateAnalyzer(semanticModel, context.CancellationToken)
+                .ComputeRefactoringsAsync(context).ConfigureAwait(false);
+        }
+
+        protected abstract IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken);
+
+        protected interface IAnalyzer
+        {
+            Task ComputeRefactoringsAsync(CodeRefactoringContext context);
+        }
+
+        protected abstract class AnalyzerBase<TRefactor, TSource, TDestination> : IAnalyzer
+            where TRefactor : SyntaxNode
+            where TSource : SyntaxNode
+            where TDestination : SyntaxNode
+        {
+            protected readonly SemanticModel _semanticModel;
+            protected readonly CancellationToken _cancellationToken;
+
+            public AnalyzerBase(SemanticModel semanticModel, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _cancellationToken = cancellationToken;
+            }
+
+            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+            {
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var refactor = FindNodeToRefactor(root, context);
+                if (refactor == null)
+                {
+                    return;
+                }
+
+                if (refactor.ContainsDiagnostics)
+                {
+                    return;
+                }
+
+                if (TryConvert(refactor, out var result))
+                {
+                    if (!result.IsValid())
```

Comment:
seems weird that TryConvert would return 'true', but then say "i'm not valid".  Why have two ways to say you've succeeded.